### PR TITLE
fix delete product uses broad exception

### LINKED ISSUE #37: Use specific exception on delete_product
`delete_product` is using broad exception (`except Exception`). This will make us ignore all failures there, which is seldom what we're after.

The exception needs to change so we ignore only the specific exception that is raised when we try to delete a non-existing element.

### PR DESCRIPTION
**Pull request by thetaxas2468**
Here we are fixing the delete product function
that can cause an exception of not exist product
so we catch it and raise it after deleting it(after the save) and if it does exist it deletes it and
returns true.
- fixed #37

### CODE REVIEW

File: product/models.py [hunk 1]
```diff
@@ -36,6 +36,8 @@ def save_product(self):
     @staticmethod
     def delete_product(prd_qr):
         try:
-            Product.filter_qr(prd_qr).delete()
-        except Exception:
-            pass
+            prod = Product.objects.get(qr_code=prd_qr)
+            prod.delete()
+        except Product.DoesNotExist as e:
```
Comment (yftacherzog) [inline product/models.py:41]: With the current implementation, I'm not sure I understand the idea behind having this function anymore (that also relates to what Barak mentioned in class on one of the other team's presentations).

What's the benefit of using this static method comparing to the built-in `delete` member method?
It just adds code that, in the best case, does exactly what the built in code does but requires you to maintain additional code.

It's better to use preexisting code (which is someone else's problem), especially if it comes from a library you already import on your codebase, comparing to start maintaining additional code that does the same.

File: product/tests/test_product.py [hunk 2]
```diff
@@ -28,6 +28,12 @@ def test_save_product_and_delete_product(self, saved_product0):
         assert saved_product0 in Product.objects.all()
         Product.delete_product(saved_product0.qr_code)
         assert saved_product0 not in Product.objects.all()
+        res = True
+        try:
+            Product.delete_product("NOTEXISTEDTEST")
+        except Product.DoesNotExist:
+            res = False
+        assert not res
```
Comment (yftacherzog) [inline product/tests/test_product.py:36]: We don't typically do try-except inside tests. Instead, we assert that the exception was raised with context manager `with pytest.raises(SomeException) as e`.

See examples here: https://stackoverflow.com/a/23514853

**Comment by yftacherzog** [comment]
@thetaxas2468 Regarding the commit title, you might have more commits in the future for fixing stuff on delete product.

For that reason, you should be as informative as possible on the commit title, while still keeping the limit of 50 characters you decided upon (your contributing guide says 50 chars for title, 72 chars for body).

In this case, I would go with something like: `fix delete product uses broad exception`.

File: product/models.py (hunk 1 above)
Comment (thetaxas2468) [reply product/models.py:41]: Fixed

File: product/tests/test_product.py (hunk 2 above)
Comment (thetaxas2468) [reply product/tests/test_product.py:36]: Fixed

**Comment by thetaxas2468** [comment]
> @thetaxas2468 Regarding the commit title, you might have more commits in the future for fixing stuff on delete product.
> 
> For that reason, you should be as informative as possible on the commit title, while still keeping the limit of 50 characters you decided upon (your contributing guide says 50 chars for title, 72 chars for body).
> 
> In this case, I would go with something like: `fix delete product uses broad exception`.

fixed

File: product/models.py [hunk 3]
```diff
@@ -29,13 +29,9 @@ def save_product(self):
...
             raise e
         return self
 
-    @staticmethod
-    def delete_product(prd_qr):
-        try:
-            Product.filter_qr(prd_qr).delete()
-        except Exception:
-            pass
+    def __str__(self) -> str:
```
Comment (Omeramsc) [inline product/models.py:36]: Is that relevant to this PR?

Comment (Omeramsc) [inline product/models.py:36]: Is this function relevant to this PR?

File: product/tests/test_product.py [hunk 4]
```diff
@@ -16,20 +17,34 @@ def product0():
...
+@pytest.fixture
+def saved_product1(product1):
+    product1.save_product()
+    return product1
+
+
 class TestProductModel:
     @pytest.mark.django_db()
-    def test_save_product_and_delete_product(self, saved_product0):
+    def test_save_product_and_delete_product(self, saved_product0, saved_product1):
```
Comment (Omeramsc) [inline product/tests/test_product.py:39]: I'll let it slide, but do note that tests usually should check for one thing (e.g: a test for saving a product and a test for deleting a product). then, if a test failed, you can pinpoint where the issue lies more easily.

File: product/models.py (hunk 3 above)
Comment (thetaxas2468) [reply product/models.py:36]: It is a super function that i overrides for debuging,it can be in any pr because each object has it

File: product/tests/test_product.py (hunk 4 above)
Comment (yftacherzog) [reply product/tests/test_product.py:39]: This was my suggestion in this case. @thetaxas2468 proposed separate tests, but I suggested combining them for this purpose as it made sure that we add a product that doesn't exist, and that we delete a product that does exist.

Comment (thetaxas2468) [reply product/tests/test_product.py:39]: I have done that before

File: product/tests/test_product.py [hunk 5]
```diff
@@ -16,20 +17,34 @@ def product0():
...
         assert saved_product0 in Product.objects.all()
-        Product.delete_product(saved_product0.qr_code)
+        saved_product0.delete()
         assert saved_product0 not in Product.objects.all()
-        saved_product0.save_product()
-        assert saved_product0 in Product.objects.all()
+        with pytest.raises(Product.DoesNotExist):
+            test_doesnotexist = Product("DOESNOTEXIST", "TEST", "TEST")
+            test_doesnotexist.delete()
+            raise Product.DoesNotExist()
```
Comment (yftacherzog) [inline product/tests/test_product.py:46]: The code should do the raising rather than the test.

Comment (thetaxas2468) [reply product/tests/test_product.py:46]: I dont understand,can you explain?

Comment (yftacherzog) [reply product/tests/test_product.py:46]: We use that context manager (`with pytest.raises(SomeExceptionClass)`) when we test a scenario in which the code that we test should fail and raised an exception of that class. The context manager asserts that the exception was raised.

If you raise the exception inside the test, then your test will always pass. It's the equivalent of writing `assert True` on your test.

Does the built-in `delete` method raises something when deleting a non-existing product? If not, then you don't need that context manager.

Comment (thetaxas2468) [reply product/tests/test_product.py:46]: The code first creates an instance of the Product class called test_doesnotexist with the name "DOESNOTEXIST", and then calls the delete() method on that instance. If the delete() method raises a Product.DoesNotExist exception, the code will continue to run. If the exception is not raised, the pytest library will raise a Failed error

Comment (yftacherzog) [reply product/tests/test_product.py:46]: How is that helping us? that last line makes the test to always pass. You could put any code between the `with` statement and the `raise` statement, and the test will pass.

Comment (thetaxas2468) [reply product/tests/test_product.py:46]: Sorry,buy i really cant understand what is wrong and why it is always passing as you are saying

Comment (yftacherzog) [reply product/tests/test_product.py:46]: The idea of having that context manager is for testing that the exception was indeed raised. I can delete the 2 lines calling the code being tested and the test will still pass, so what does it test?
```
        with pytest.raises(Product.DoesNotExist):
            raise Product.DoesNotExist()
```

Comment (thetaxas2468) [reply product/tests/test_product.py:46]: try:
            test_doesnotexist = Product("DOESNOTEXIST", "TEST", "TEST")
            test_doesnotexist.delete_product()
        except Product.DoesNotExist:
            assert False
    def delete_product(self):
        Product.objects.get(qr_code=self.qr_code).delete()

this code raise Product Does Not Exist,i have converted that to context manager and it passes

Comment (thetaxas2468) [reply product/tests/test_product.py]: regular delete does not raise the does not exist as i tried,if you want it to be raised first you need to get the row with the specific qr code and then delete it if it already exist

Comment (yftacherzog) [reply product/tests/test_product.py]: When is the `DoesNotExist` exception being raised? When you run the test, you should know whether it need to be raised or not, and assert accordingly.

Comment (thetaxas2468) [reply product/tests/test_product.py]: In the new code the doesnotexist is being raised that is why the test have completed,other wise the test should fail i dont need to assert nothing i guess,since the product does not exist raised assert true,other wise assert false or exception

Comment (yftacherzog) [reply product/tests/test_product.py]: I lost you.

Do you want your code to raise an exception when you delete? Are you currently just adding a new method `delete_product` so it will raise an exception?

I honestly don't know if you should have it raise an exception or not. It's up to you to decide. I'd say:
- If you know that you want it to raise an exception, then fine.
- If you don't know that you need that, then don't include that code (you could add it later on if needed).
- If you don't add the code that raises an exception, then don't test that the an exception is being raised.

Comment (thetaxas2468) [reply product/tests/test_product.py]: - I need that code and yes i wanted that exception to be raised and the exception is tested we can see the test passing
- I dont want a situation that you are ready to delete a non existed prouduct

Comment (yftacherzog) [reply product/tests/test_product.py]: ok. I'm fine with that I think. I'm going to add some comments on the current code (sorry about that).

File: product/tests/test_product.py [hunk 6]
```diff
@@ -16,20 +17,33 @@ def product0():
...
+    def test_save_product_and_delete_product(self, saved_product0, saved_product1):
         assert saved_product0 in Product.objects.all()
-        Product.delete_product(saved_product0.qr_code)
+        saved_product0.delete()
         assert saved_product0 not in Product.objects.all()
-        saved_product0.save_product()
-        assert saved_product0 in Product.objects.all()
+        with pytest.raises(Product.DoesNotExist):
+            test_doesnotexist = Product("DOESNOTEXIST", "TEST", "TEST")
+            test_doesnotexist.delete_product()
```
Comment (yftacherzog) [inline product/tests/test_product.py:45]: This should be on a test function of its own, as it's not related to the delete-save cycle that we're testing, and while you do that, make the context manager include just the line that should raise the exception.

```
        test_doesnotexist = Product("DOESNOTEXIST", "TEST", "TEST")
        with pytest.raises(Product.DoesNotExist):
            test_doesnotexist.delete_product()
```

File: product/tests/test_product.py [hunk 7]
```diff
@@ -16,20 +17,33 @@ def product0():
...
+        assert saved_product1 in Product.objects.all()
```
Comment (yftacherzog) [inline product/tests/test_product.py:46]: We're again testing the fixture here, and I don't see a reason to do this.

File: product/tests/test_product.py (hunk 6 above)
Comment (thetaxas2468) [reply product/tests/test_product.py:45]: fixed

File: product/tests/test_product.py (hunk 7 above)
Comment (thetaxas2468) [reply product/tests/test_product.py:46]: fixed

File: product/tests/test_product.py [hunk 8]
```diff
@@ -24,12 +25,19 @@ def saved_product0(product0):
...
-        Product.delete_product(saved_product0.qr_code)
+        saved_product0.delete()
         assert saved_product0 not in Product.objects.all()
-        saved_product0.save_product()
-        assert saved_product0 in Product.objects.all()
+        temp.save()
+        assert temp in Product.objects.all()
+
+    @pytest.mark.django_db()
+    def test_delete_product(self):
```
Comment (yftacherzog) [inline product/tests/test_product.py:37]:
```suggestion
    def test_delete_non_existing_product(self):
```

File: product/tests/test_product.py [hunk 9]
```diff
@@ -24,12 +25,19 @@ def saved_product0(product0):
 
 class TestProductModel:
     @pytest.mark.django_db()
-    def test_save_product_and_delete_product(self, saved_product0):
+    def test_save_product(self, saved_product0):
+        temp = Product(saved_product0.qr_code, saved_product0.product_name, saved_product0.description)
         assert saved_product0 in Product.objects.all()
-        Product.delete_product(saved_product0.qr_code)
+        saved_product0.delete()
```
Comment (yftacherzog) [inline product/tests/test_product.py:31]: This should be `delete_product` if we're back to use it.

File: product/models.py (hunk 3 above)
Comment (yftacherzog) [reply product/models.py:39]: I agree that this should not be a part on this PR. If you want it added then you need a different PR for that (which also includes tests for that function).

File: product/tests/test_product.py (hunk 9 above)
Comment (thetaxas2468) [reply product/tests/test_product.py:31]: fixed

File: product/tests/test_product.py [hunk 10]
```diff
@@ -2,6 +2,7 @@
 from product.models import Product
 from django.core.exceptions import ValidationError
 
+QR1 = "AAAA6MbdiNbXprNEnHfpcGWFp1CMF8XY"
```
Comment (yftacherzog) [inline product/tests/test_product.py:5]: not being used

Comment (thetaxas2468) [reply product/tests/test_product.py:5]: Right,forgot it thanks.fixed.

File: product/models.py (hunk 3 above)
Comment (Omeramsc) [reply product/models.py:39]: Fix the above and we can merge this PR.

Comment (thetaxas2468) [reply product/models.py]: fixed

File: product/tests/test_product.py [hunk 11]
```diff
@@ -24,12 +24,19 @@ def saved_product0(product0):
...
+    def test_save_delete_product(self, saved_product0):
+        temp = Product(saved_product0.qr_code, saved_product0.product_name, saved_product0.description)
         assert saved_product0 in Product.objects.all()
-        Product.delete_product(saved_product0.qr_code)
+        saved_product0.delete_product()
         assert saved_product0 not in Product.objects.all()
-        saved_product0.save_product()
-        assert saved_product0 in Product.objects.all()
+        temp.save()
+        assert temp in Product.objects.all()
```
Comment (pazMenachem) [inline product/tests/test_product.py:33]: wondering whats the difference between temp and saved product0? why the need for temp?

**Review by pazMenachem — COMMENTED** [review]
other than the comment, rest seems fine, good example how short PR could be effective.

**Review by Omeramsc — APPROVED** [review]
LGTM
